File: cpp/cpp_1-9/3.cc
/*
 * C++
 * leetcode 3
 * url: https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/
 * 用p一次遍历
 * 一个int start用来记录当前字符串起点
 * 每次p判断流程：
 * 		用i遍历start到p-1，s[i]是否与s[p]相等
 *		* 存在相等的则用p-start与max_global相比，更大的话则刷新max_global，且更新start = 相等的i +1
 *		* 不存在相等的则可以不做任何事
 * 最后还需要用p-start刷一次max_global
 */

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
    	int size = s.size(), max_global =1;
    	if(size==0)
    		return 0;
    	int start = 0,p =1;
    	while(p<size){
    		for(int i =start; i<p; i++){
    			if(s[i] ==s[p]){
    				if(p-start>max_global)
    					max_global = p-start;
    				start =i+1;
    			}
    		}
    		p++;
    	}
		if(p-start>max_global)
			max_global = p-start;
		return max_global;
    }
};

File: Go/Go_1-9/3.go
// Go
// leetcode 3
// https://leetcode.cn/problems/longest-substring-without-repeating-characters/

func lengthOfLongestSubstring(s string) int {
	vis, n, res := make([]bool, 128), len(s), 0
	for i, j := 0, 0; j < n; j++ {
		idx_j := int(s[j])
		if vis[idx_j] {
			for int(s[i]) != idx_j {
				vis[int(s[i])] = false
				i++
			}
			i++
		} else {
			vis[idx_j] = true
		}
		if j-i+1 > res {
			res = j - i + 1
		}
	}
	return res
}